File: backend/app/schemas/friends.py
from pydantic import BaseModel
from typing import Optional
from uuid import UUID
from datetime import datetime


class FriendRequestCreate(BaseModel):
    addressee_id: str


class FriendshipResponse(BaseModel):
    id: UUID
    requester_id: UUID
    addressee_id: UUID
    status: str
    is_favorite: bool
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class FriendDetail(BaseModel):
    id: str
    friendship_id: Optional[str] = None
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    bio: Optional[str]
    is_verified: bool
    mutual_friends_count: int = 0
    is_close_friend: bool = False
    model_config = {"from_attributes": True}


class FollowResponse(BaseModel):
    id: UUID
    follower_id: UUID
    following_id: UUID
    created_at: datetime
    model_config = {"from_attributes": True}


class FollowUserDetail(BaseModel):
    id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    bio: Optional[str]
    is_verified: bool
    is_friend: bool = False
    mutual_friends_count: int = 0
    model_config = {"from_attributes": True}


class FriendshipStatusResponse(BaseModel):
    status: Optional[str]
    is_requester: bool
    is_favorite: bool
    is_close_friend: bool
    is_following: bool = False
    is_followed_by: bool = False


class FollowRequestResponse(BaseModel):
    id: UUID
    requester_id: UUID
    target_id: UUID
    status: str
    created_at: datetime
    model_config = {"from_attributes": True}


class FollowRequestDetail(BaseModel):
    id: str
    request_id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    bio: Optional[str]
    is_verified: bool
    model_config = {"from_attributes": True}


class CloseFriendResponse(BaseModel):
    id: UUID
    user_id: UUID
    friend_id: UUID
    created_at: datetime
    model_config = {"from_attributes": True}


class CloseFriendDetail(BaseModel):
    id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    bio: Optional[str]
    is_verified: bool
    added_at: datetime
    model_config = {"from_attributes": True}


class FavoriteUpdate(BaseModel):
    is_favorite: bool


FavoriteCloseUpdate = FavoriteUpdate


class MuteUpdate(BaseModel):
    mute_posts: bool = True
    mute_stories: bool = True
    mute_notes: bool = True
    mute_notifications: bool = True


class MuteResponse(BaseModel):
    id: UUID
    user_id: UUID
    muted_user_id: UUID
    mute_posts: bool
    mute_stories: bool
    mute_notes: bool
    mute_notifications: bool
    created_at: datetime
    model_config = {"from_attributes": True}


class MuteDetail(BaseModel):
    id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    mute_posts: bool
    mute_stories: bool
    mute_notes: bool
    mute_notifications: bool
    model_config = {"from_attributes": True}


class RestrictResponse(BaseModel):
    id: UUID
    user_id: UUID
    restricted_user_id: UUID
    created_at: datetime
    model_config = {"from_attributes": True}


class RestrictDetail(BaseModel):
    id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    restricted_at: datetime
    model_config = {"from_attributes": True}


class BlockResponse(BaseModel):
    id: str
    blocked_user_id: str
    blocked_at: datetime


class BlockDetail(BaseModel):
    id: str
    full_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    blocked_at: datetime
    model_config = {"from_attributes": True}


class UserCounts(BaseModel):
    friends: int
    followers: int
    following: int
    close_friends: int
    pending_friend_requests: int = 0
    pending_follow_requests: int = 0


class RelationshipSummary(BaseModel):
    are_friends: bool
    is_close_friend: bool
    is_following: bool
    is_followed_by: bool
    are_blocked: bool
    is_muted: bool
    is_restricted: bool
    mutual_friends_count: int = 0

File: frontend/src/types/friends.ts
export interface FriendRequest {
  addressee_id: string;
}

export interface Friendship {
  id: string;
  requester_id: string;
  addressee_id: string;
  status: "pending" | "accepted" | "rejected";
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface FriendDetail {
  id: string;
  friendship_id: string | null;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  bio: string | null;
  is_verified: boolean;
  mutual_friends_count: number;
  is_close_friend: boolean;
}

export interface FollowUser {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  bio: string | null;
  is_verified: boolean;
  is_friend: boolean;
  mutual_friends_count: number;
}

export interface FriendshipStatus {
  status: "pending" | "accepted" | "rejected" | null;
  is_requester: boolean;
  is_favorite: boolean;
  is_close_friend: boolean;
  is_following: boolean;
  is_followed_by: boolean;
}

export interface FriendCounts {
  friends: number;
  followers: number;
  following: number;
  close_friends: number;
  pending_friend_requests: number;
  pending_follow_requests: number;
}

export interface FavoriteUpdate {
  is_favorite: boolean;
}

export interface CloseFriend {
  id: string;
  user_id: string;
  friend_id: string;
  created_at: string;
}

export interface CloseFriendDetail {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  bio: string | null;
  is_verified: boolean;
  added_at: string;
}

export interface FollowRequest {
  id: string;
  requester_id: string;
  target_id: string;
  status: "pending" | "accepted" | "rejected";
  created_at: string;
}

export interface FollowRequestDetail {
  id: string;
  request_id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  bio: string | null;
  is_verified: boolean;
}

export interface Mute {
  id: string;
  user_id: string;
  muted_user_id: string;
  mute_posts: boolean;
  mute_stories: boolean;
  mute_notes: boolean;
  mute_notifications: boolean;
  created_at: string;
}

export interface MuteDetail {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  mute_posts: boolean;
  mute_stories: boolean;
  mute_notes: boolean;
  mute_notifications: boolean;
}

export interface MuteUpdate {
  mute_posts?: boolean;
  mute_stories?: boolean;
  mute_notes?: boolean;
  mute_notifications?: boolean;
}

export interface Restrict {
  id: string;
  user_id: string;
  restricted_user_id: string;
  created_at: string;
}

export interface RestrictDetail {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  restricted_at: string;
}

export interface Block {
  id: string;
  blocked_user_id: string;
  blocked_at: string;
}

export interface BlockDetail {
  id: string;
  full_name: string | null;
  username: string | null;
  avatar_url: string | null;
  blocked_at: string;
}

export interface RelationshipSummary {
  are_friends: boolean;
  is_close_friend: boolean;
  is_following: boolean;
  is_followed_by: boolean;
  are_blocked: boolean;
  is_muted: boolean;
  is_restricted: boolean;
  mutual_friends_count: number;
}

export type FriendsTab = "all" | "suggestions" | "pending" | "followers" | "following" | "close";
